cub/util_type.cuh

```diff
@@ -273,9 +274,47 @@ struct NullType
 template <int A>
 struct Int2Type
 {
-   enum {VALUE = A};
+    enum {VALUE = A};
 };
 
+/**
+ * \brief Allows algorithms that take a value as input to take a future value that is not computed yet at launch time.
+ *
+ * Note that it is user's responsibility to ensure that the result will be ready before use via external synchronization
+ * or stream-ordering dependencies.
+ */
+template <typename T, typename IterT = T*>
+struct FutureValue
+{
+    using value_type = T;
+    using iterator_type = IterT;
+    IterT iter;
+    __host__ __device__ __forceinline__ FutureValue(IterT iter):iter(iter) {}
+    __host__ __device__ __forceinline__ operator T() {
+        return *iter;
+    }
+};
+
+template <typename T, typename IterT = T*>
```

Comment:
I didn't think of this in our initial discussion, but let's move this into the `cub::detail::` namespace, since users shouldn't directly work with it. It will also need a documentation string explaining why it exists.

Reply:
fixed